src/gt4py/next/iterator/transforms/constant_folding.py

```diff
@@ -6,52 +6,245 @@
 # Please, refer to the LICENSE file in the root directory.
 # SPDX-License-Identifier: BSD-3-Clause
 
-from gt4py.eve import NodeTranslator, PreserveLocationVisitor
+from __future__ import annotations
+
+import dataclasses
+import enum
+import functools
+import operator
+from typing import Optional
+
+from gt4py import eve
 from gt4py.next.iterator import builtins, embedded, ir
-from gt4py.next.iterator.ir_utils import ir_makers as im
+from gt4py.next.iterator.ir_utils import common_pattern_matcher as cpm, ir_makers as im
+from gt4py.next.iterator.transforms import fixed_point_transformation
+
+
+@dataclasses.dataclass(frozen=True, kw_only=True)
+class ConstantFolding(
+    fixed_point_transformation.FixedPointTransformation, eve.PreserveLocationVisitor
+):
+    PRESERVED_ANNEX_ATTRS = (
+        "type",
+        "domain",
+    )
+
+    class Transformation(enum.Flag):
+        # e.g. `literal + symref` -> `symref + literal` and
+        # `literal + funcall` -> `funcall + literal` and
+        # `symref + funcall` -> `funcall + symref`
+        CANONICALIZE_FUNCALL_SYMREF_LITERAL = enum.auto()
+
+        # `minus(arg0, arg1) -> plus(im.call("neg")(arg1), arg0)`
+        CANONICALIZE_MINUS = enum.auto()
+
+        # `maximum(im.call(...)(), maximum(...))` -> `maximum(maximum(...), im.call(...)())`
+        CANONICALIZE_MIN_MAX = enum.auto()
+
+        # `im.call(...)(im.tuple_get(...), im.plus(...)))` -> `im.call(...)( im.plus(...)), im.tuple_get(...))`
+        CANONICALIZE_TUPLE_GET_PLUS = enum.auto()
+
+        # `(sym + 1) + 1` -> `sym + 2`
+        FOLD_FUNCALL_LITERAL = enum.auto()
+
+        # `maximum(maximum(sym, 1), sym)` -> `maximum(sym, 1)`
```

Comment:
```suggestion
        # `maximum(maximum(a, 1), b)` -> `maximum(maximum(a, b), 1)`
```
and for plus, multiplies.